### DISCUSSION TITLE
Path filtering on required pull request checks

### DISCUSSION THREAD

**Discussion by ConnorGraham** [community/community#26857] [Actions]
<p>We have a monorepo holding multiple microservices. On a pull request, we want to only run unit tests on a service that has been changed within the PR. We accomplish this with a workflow syntax like this:</p>
<pre><code>on:
  pull_request:
    paths:
      - 'serviceA/*'
</code></pre>
<p>However, when we mark the branch protection check as required then we can’t merge any pull requests that do not include changes to serviceA. Is there a way we can work around this?</p>

**Comment by weide-zhou** [comment]
<p>Hi <a class="mention" href="/connorgraham">@connorgraham</a> ,</p>
<p>I can repro the same on my side, the required check run will stay at ‘Expected- Waiting…’ status.</p>
<p>You can remove paths setting for pull_request event, and remove checkrun from branch protection rule. Instead, check the file commit info, execute different jobs according to the file path. If it comes from service A, execute job1, if it comes from other file, execute job2…etc, it will skip unneccessary jobs and only one check run in pull request.</p>
<p>Please check the <a href="https://github.community/t/get-list-of-files-on-pull-request-merge/17226">link</a> about how to get the file name(path).</p>
<p>Or you can create another yaml file with different paths setting, cover the files changes out of service A.</p>
<p>Thanks</p>

**Comment by ConnorGraham** [comment, marked as answer]
<p>Thanks <a class="mention" href="/weide-zhou">@weide-zhou</a> </p>
<p>I solved it using the github actions script:</p>
<pre><code>on: [pull_request]

jobs:
   test:
    name: Test
    runs-on: ubuntu-latest
    steps:
      - name: Check if service has changed
        uses: actions/github-script@0.9.0
        id: service-changed
        with:
          result-encoding: string
          script: |
            const result = await github.pulls.listFiles({
              owner: context.payload.repository.owner.login,
              repo: context.payload.repository.name,
              pull_number: context.payload.number,
              per_page: 100
            })
            const serviceChanged = result.data.filter(f =&gt; f.filename.startsWith("serviceA/")).length &gt; 0
            console.log(serviceChanged)
            return serviceChanged

      - uses: actions/checkout@v2
        if: ${{ steps.service-changed.outputs.result == 'true' }}
</code></pre>

**Comment by fkirc** [comment]
<p>I published an Action that solves all those required-check-problems: <a href="https://github.com/marketplace/actions/skip-duplicate-actions" rel="noopener nofollow ugc">https://github.com/marketplace/actions/skip-duplicate-actions</a></p>
<p>This is similar to GitHub’s path filter, but the underlying algorithm is smarter.<br>
You can give it a try if this is still an issue.</p>

**Comment by piotrekkr** [comment]
Another "solution" found here https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/defining-the-mergeability-of-pull-requests/troubleshooting-required-status-checks#handling-skipped-but-required-checks

**Comment by mrchief** [comment]
You can add a `no-op.yml` (or any name) that adds a dummy exit code for the required job:

```
# Workaround for required checks when file patterns don't match.

name: No Op
on:
  pull_request:

jobs:
  <name of the required job>:
    runs-on: ubuntu-latest
    steps:
      - name: Do nothing
        run: exit 0
```

The `<name of the required job>` needs to be the same as the required check, so e.g., if your required check is called `lint-pull-request` then 

```
...

jobs:
  lint-pull-request:
...
```

This will short circuit the required check when it is skipped and if the the required fails, this will not overwrite that so the PR will still be in a failed state.

**Comment by Maxim-Durand** [comment]
I tried to use [skip-duplicate-actions](https://github.com/marketplace/actions/skip-duplicate-actions) but I think it's currently bugged and won't skip the jobs, instead I ended up using [changed-files](https://github.com/tj-actions/changed-files).
I won't post an example here because [changed-files](https://github.com/tj-actions/changed-files) provides already plenty of examples in their doc, for instance have a look at https://github.com/tj-actions/changed-files#on-pull_request-

**Comment by schammah** [comment]
solution works but it is very inefficient, lots of status checks running for nothing.
any feature requests for bitcoin to support this?
